demos/text-generation/llm.js

```diff
@@ -0,0 +1,321 @@
+                t.dispose();
+            }
+        }
+
+        this.feed = {};
+        if (this.provider == "webnn") {
+            // init kv cache ml-tensor
+            const kvDesc = { dataType: this.dtype, shape: this.kvDims };
+            const ortKvDesc = { dataType: this.dtype, dims: this.kvDims };
+            const inputMlTensor = await this.mlContext.createTensor(kvDesc);
+            for (let i = 0; i < this.numLayers; ++i) {
+                this.feed[`past_key_values.${i}.key`] = ort.Tensor.fromMLTensor(inputMlTensor, ortKvDesc);
+                this.feed[`past_key_values.${i}.value`] = ort.Tensor.fromMLTensor(inputMlTensor, ortKvDesc);
+            }
+        } else {
+            const kvNumElements = product(this.kvDims);
+            const empty = this.dtype === "float16" ? new Float16Array(kvNumElements) : new Float32Array(kvNumElements);
+            for (let i = 0; i < this.numLayers; ++i) {
+                this.feed[`past_key_values.${i}.key`] = new ort.Tensor(this.dtype, empty, this.kvDims);
+                this.feed[`past_key_values.${i}.value`] = new ort.Tensor(this.dtype, empty, this.kvDims);
+            }
+        }
+    }
+
+    // update key value cache
+    updateKvCache(outputs) {
+        for (const name in outputs) {
+            if (name.includes("present.")) {
+                let newName = name.replace(name.split(".")[0], "past_key_values");
+                const t = this.feed[newName];
+                // dispose previous tensors
+                if (t.location === "gpu-buffer" || t.location == "ml-tensor") {
+                    // ml-tensor in first feed has no dispose method
+                    if (t.disposer == undefined && t.location == "ml-tensor") {
+                        t.mlTensor.destroy();
+                    } else {
+                        t.dispose();
+                    }
+                }
+
+                this.feed[newName] = outputs[name];
+            }
+        }
+    }
+
+    // padding input array with 0
+    paddingInput(originInput, maxLength, reverse = false) {
+        let input = originInput.slice();
+        if (input.length >= maxLength) return input.slice(0, maxLength);
+        const paddingLength = maxLength - input.length;
+        const padding = Array.from({ length: paddingLength }, () => 0n);
+        if (reverse) {
+            padding.push(...input);
+            return padding;
+        } else {
+            input.push(...padding);
+            return input;
+        }
+    }
+
+    // tell generate to stop()
+    abort() {
+        this.stop = true;
+    }
+
+    // poor mens argmax
+    argmax(t, seqLen = 1) {
+        let arr = t.cpuData;
+        if (t.type == "float16" && !isFloat16ArrayAvailable) {
+            throw new Error("Float16Array is not available on this browser, try to use newer version");
+        }
+
+        let start = t.dims[2] * (seqLen - 1);
+        let max = arr[start];
+        let maxidx = 0;
```

Comment:
`maxIdx` to follow consistent camelCase (or better yet, `maxIndex`)
Consider `sequenceLength` too.